server/src/session.ts

```diff
@@ -91,16 +92,38 @@ export class Session {
       case ts.server.ProjectsUpdatedInBackgroundEvent:
         // ProjectsUpdatedInBackgroundEvent is sent whenever diagnostics are
         // requested via project.refreshDiagnostics()
-        this.refreshDiagnostics(event.data.openFiles);
+        this.triggerDiagnostics(event.data.openFiles);
         break;
     }
   }
 
   /**
-   * Retrieve Angular diagnostics for the specified `openFiles`.
+   * Retrieve Angular diagnostics for the specified `openFiles` after a specific
+   * `delay`, or drop the request if there's already a pending one.
    * @param openFiles
+   * @param delay time to wait before sending request (milliseconds)
    */
-  private refreshDiagnostics(openFiles: string[]) {
+  private triggerDiagnostics(openFiles: string[], delay: number = 200) {
+    // Do not immediately send a diagnostics request. Throttle the number of
+    // requests to at most five per second. (200ms delay, consistent with
+    // TypeScript. See
+    // https://github.com/microsoft/vscode/blob/7b944a16f52843b44cede123dd43ae36c0405dfd/extensions/typescript-language-features/src/features/bufferSyncSupport.ts#L493)
+    // A request is dropped if there's already a pending one.
+    if (this.hasPendingDiagnosticsRequest) {
+      return;
+    }
+    this.hasPendingDiagnosticsRequest = true;
+    setTimeout(() => {
+      this.hasPendingDiagnosticsRequest = false;
+      this.sendPendingDiagnostics(openFiles);
```

Comment:
Regardless of what we do, I think the diagnostic should be sent immediately and only `hasPendingDiagnosticsRequest` is updated after 200ms. Otherwise we are waiting for even a diagnostic to be triggered.